cmds/mem.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Phoenix-RTOS
+ *
+ * MEMory read/write tool
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/lib.h>
+#include <syspage.h>
+#include "cmd.h"
+
+
+static void cmd_memInfo(void)
+{
+	lib_printf("Reads or writes values from/to a range of memory addresses");
+}
+
+
+static void cmd_memUsage(void)
+{
+	lib_printf(
+		"Usage: mem [options].. address\n"
+		"\t-F        Disable validation of memory (map) regions\n"
+		"\t-a align  Memory alignment (default=%u)\n"
+		"\t-s size   Size of write/read (1,2,4,8)\n"
+		"\t-c count  Number of repetitions with address step of (size)\n"
+		"\t-w value  Write (value) to memory address\n"
+		"\t-r        Read memory from address\n",
+		(unsigned int)sizeof(addr_t));
+}
+
+
+static int region_validator(addr_t addr, size_t size, unsigned int attr)
+{
+	unsigned int mapAttr = 0uL;
+
+	if (syspage_mapRangeCheck(addr, addr + size, &mapAttr) == 0) {
+		return 0;
+	}
+
+	return ((mapAttr & attr) == attr) ? 1 : 0;
+}
+
+
+static int memr(addr_t addr, size_t size, int forceValid)
+{
+	union {
+		u64 val64;
+		u32 val32;
+		u16 val16;
+		u8 val8;
+	} u = { .val64 = 0uLL };
+
+	if (sizeof(addr_t) <= 8uLL) {
+		lib_printf("read 0x%08x = ", addr);
+	}
+	else {
+		lib_printf("read 0x%016llx = ", (u64)addr);
+	}
+
+	if ((forceValid != 0) || (region_validator(addr, size, mAttrRead) != 0)) {
+		switch (size) {
+			case sizeof(u.val8):
+				hal_cpuDataMemoryBarrier();
+				u.val8 = *(u8 *)addr;
+				lib_printf("0x%02x\n", u.val8);
```

Comment:
Intermediate variable isn't really needed, but it's a matter of taste, feel free to ignore this remark

```suggestion
				lib_printf("0x%02x\n", *(u8 *)addr);
```